### ISSUE TITLE
[Tripal3]Aliases token not filled (empty)

### ISSUE THREAD

**Issue by mboudet** [tripal/tripal#790]
### System information

System info:
Tripal: 7.x-3.0
Drupal : 7.59
Postgres: 9.5.8
Php: 7.1.23

### Issue description
Hi,
I noticed that some of my automatically created aliases are malformed (missing some tokens).
This occurs for the mRNA entity type.
My alias pattern is as follows : features/[obi__organism,TAXRANK:0000005]/[obi__organism,TAXRANK:0000006]/[rdfs__type]/[data__identifier]

An expected alias would be : features/Citrus/sinensis/mRNA/PAC:18136223

But the created alias looks like this : features///mRNA/
(So there are duplicate errors)

#### Steps to reproduce
This one is complicated, because the problem only occurs if the token fields are not in cache. Depending on your setup, you will not see it occurs.

#### Issues

Looking at the code, there are actually two separate issues:
1) **In tripal.entities.api.inc**
[Here](https://github.com/tripal/tripal/blob/3606ca5cf8891a068c0dc5ebe6ac48f245252136/tripal/api/tripal.entities.api.inc#L1331)

We use the $field_ids variable to monitor the fields missing from cache, however the variable is reset in the first loop:

```
foreach ($attach_fields as $storage_type => $details) {
      $storage = $details['storage'];
      $fields = $details['fields'];
      $field_ids = array();
            foreach ($fields as $field) {
        $field_ids[$field['id']] = array($entity->id);
      }
```
So only the last loop actually keep the values. Meaning, if there are more than one storage type for the fields, information is lost. (In my case, the [rdfs__type] token storage is tripal_no_storage, and the other tokens are field_chado_storage.
This is easily fixed by moving the $field_ids = array(); before the first loop.

2) **In tripal_field_storage_load**

This one is a bit more complicated, but basically, we ignore any fields which are not TripalField class. (like text..)

If we compare the hook_field_storage_load implementation in tripal and tripal_chado, we can see that the tripal implementation completely ignore non-TripalField tokens.

```
tripal_load_include_field_class($field_type);
      if (class_exists($field_type)) {
        $tfield = new $field_type($field, $instance);
        $tfield->load($entity);
      }

```
While tripal_chado implements something to take care of it
```
tripal_load_include_field_class($field_type);
        if (class_exists($field_type) and is_subclass_of($field_type, 'TripalField')) {
          $tfield = new $field_type($field, $instance);
          $tfield->load($entity, array('record' => $record));
        }
        // For text fields that were not handled by a TripalField class we
        // want to automatically expand those fields.
        else {
          if ($schema['fields'][$field_column]['type'] == 'text') {
            $record = chado_expand_var($record, 'field', "$field_table.$field_column");
            $entity->{$field_name}['und'][0]['value'] = $record->$field_column;
            // Text fields that have a text_processing == 1 setting need a
            // special 'format' element too:
            if (array(key_exists('text_processing', $instance['settings']) and
                $instance['settings']['text_processing'] == 1)) {
              // TODO: we need a way to write the format back to the
              // instance settings if the user changes it when using the form.
              $entity->{$field_name}['und'][0]['format'] = array_key_exists('format', $instance['settings']) ? $instance['settings']['format'] : 'full_html';
            }
          }
```

Which means that text tokens are simply empty in the current implementation.
To fix this, I used the implementation from tripal_chado and added it in tripal. It seems to work.

Sorry if this is not clear.

**Comment by spficklin** [comment]
I agree with the fix for issue # 1.  Thanks for spotting it!

For issue # 2...  The ``tripal_no_storage`` back-end is meant for fields that store data inside of the Drupal tables. The ``rdfs__type`` field pulls data from the bundle so that makes sense.  Any data that comes from Chado should use the ``field_chado_storage`` back-end.  This includes fields that are implemented using the TripalField class or existing Drupal fields.   If they pull data from Chado they should use that back-end.

Just to clarify... if caching is occurring for a field it comes pre-added to the entity object.  The cached values are what was pulled from the database the first time the entity is loaded for viewing on a page and cached for faster retrieval in the Drupal field tables.  If the entity has never been viewed there will be no cached fields.  Therefore, only if a field has not been cached do the lines 1327-1337 get executed:

https://github.com/tripal/tripal/blob/76a5ad6a0396c1fd4330fb77557e2c090ee3cd3d/tripal/api/tripal.entities.api.inc#L1326-L1338

So, it makes me wonder.... do you somehow have fields that are assigned to the ``tripal_no_storage`` when they should be assigned to the ``field_chado_storage``?   If they were assigned to the latter than the``tripal_chado_field_storage_load`` function would be called for those fields and the data should get loaded via the second block of code you reference in your post above.

**Comment by mboudet** [comment]
Well, among the tokens I use, only the rdfs__type is tripal_no_storage. The two tokens I have issue with (that I have tried) are [data__identifier] and [schema__name], using the field_chado_storage backend, and text type.

The issue is that

```
module_invoke($storage['module'], 'field_storage_load', 'TripalEntity', 
        $entities, FIELD_LOAD_CURRENT, $field_ids, array());
```
 does not actually load fields which are not not TripalField class (so, text type and possibly other).

```
      $instance = field_info_instance($entity_type, $field_name, $entity->bundle);
      // Allow the creating module to alter the value if desired.  The
      // module should do this if the field has any other form elements
      // that need populationg besides the value which was set above.
      tripal_load_include_field_class($field_type);
      if (class_exists($field_type)) {
        $tfield = new $field_type($field, $instance);
        $tfield->load($entity);
      }
```

The _if_ part is not used because tripal_load_include_field_class doesn't load text type.

To put it simply, any field with $storage['module'] = 'tripal', and $field_type = 'text' are not going to be loaded if they are not already in cache. (I used the tripal_chado implementation because it had the same issue, but had a workaround for text type)

**Comment by spficklin** [comment]
Hi @mboudet .  So, in the PR #796 you added code to the hook for the `tripal_field_storage_load()` function.  Did that actually work for you?

**Comment by mboudet** [comment]
Yes it did. However I only tested for the mRNA entity, so I'm not sure if it would works for all types.

**Comment by mboudet** [comment]
Since I still have the issue, is there any better implementation to fix it? 
While #796 might have been too heavy-handed, we should still manage the case.

**Comment by spficklin** [comment]
This issue seems to have been closed automatically when I merged the PR.  Is all fixed or do we need to reopen for further work?
